model_training/ml_loss.py

```
@@ -7,28 +7,26 @@
 """
 
 import torch
+import torch.nn.functional as F
 
 
 # with mean, this is equivalent to torch.sum(diff**2) / F4f_pred.numel()
 def comparison_loss_fn(F4_pred, F4_true):
-    return torch.nn.MSELoss(reduction="mean")(F4_pred, F4_true)
+    return F.mse_loss(F4_pred, F4_true, reduction="mean")
 
 
 def unphysical_loss_fn(F4f_pred, F4f_true):
     assert F4f_true == None
 
     # enforce that number density cannot be less than zero
-    negative_density_error = torch.min(
-        F4f_pred[:, 3, :, :], torch.zeros_like(F4f_pred[:, 3, :, :])
-    )  # [sim, nu/nubar, flavor]
+    ndens = F4f_pred[:, 3, :, :]  # [sim, nu/nubar, flavor]
+    negative_density_error = torch.clamp(ndens, max=0.0)
```

total density N is recomputed with a tiny eps (torch.finfo(F4.dtype).tiny) before normalizing F4_flat to avoid divide-by-zero.

---

If the total density is zero the data is bad and we should let it fail loudly rather than quietly tweak the data to not produce an error.